### ISSUE TITLE
file handles

### ISSUE THREAD

**Issue by sonnyp** [flatpak/xdg-desktop-portal#363]
Here is my use case; I basically made a browser, I use `GtkFileChooserNative` to allow users to set the download destination. Once the file is downloaded I trigger a notification with 2 actions "Open file" and "Open folder" that use "GioAppInfo.launch_default_for_uri".

See https://github.com/sonnyp/Tangram/pull/54

Unfortunately with flatpak that won't work because `GtkFileChooserNative.get_uri` returns something like `file:///run/user/1000/doc/90e34d70/MyFile.txt`.

I know this description is very Gtk oriented but I couldn't find any way to achieve that in the spec anyway.

Basically what I need is some kind of handle I could pass to various APIs that open trusted processes ?

**Comment by matthiasclasen** [comment]
I'm not entirely clear what you are trying to achieve, and what the problem is.

Your app is a flatpak that downloads files ?

And it tries to open the files using other flatpak'ed apps ?

**Comment by sonnyp** [comment]
> Your app is a flatpak that downloads files ?

Yes.

> And it tries to open the files using other flatpak'ed apps ?

Exactly. 

And I'd like to do so without any filesystem permission, just using portals.

**Comment by matthiasclasen** [comment]
If you use the file chooser portal to save a file outside the sandbox, then yes,the location you get will be something like file:///run/user/1000/doc/90e34d70/MyFile.txt, because we use the document portal to make the user-chosen location available for writing to your sandboxed app.

I'm not sure I understand where the problem is.

**Comment by sonnyp** [comment]
> If you use the file chooser portal to save a file outside the sandbox, then yes,the location you get will be something like file:///run/user/1000/doc/90e34d70/MyFile.txt, because we use the document portal to make the user-chosen location available for writing to your sandboxed app.

Yep makes sense.

> The canonical way to implement 'Open file' inside a flatpak app would be to use the OpenURI portal.

Yep but that doesn't work with a "sandbox" `file://` URI. Right?

> I'm not sure I understand where the problem is.

Unless I'm missing/misunderstanding something I guess it's more of a feature request than a problem.

I would like to be able to save a file **and** open it using portals. Which implies that

* my app doesn't know the real path of the file
* OpenURI portal (or equivalent) would somehow be able to map that "sandbox" `file:` URI to the real host path URI.

**Comment by matthiasclasen** [comment]
> > The canonical way to implement 'Open file' inside a flatpak app would be to use the OpenURI portal.
> 
> But that doesn't work with a "sandbox" `file://` URI. Right?
>

That is true. We are not allowing file:// uris in OpenURI, since it is too easy to play tricks like opening file:///etc/passwd. But you can open the file, and pass the fd to the OpenFile method of the OpenURI portal. We require an fd as proof that you have access to the file. Behind the scenes, the portal is supposed to do all the necessary magic to make the file available to whatever app the user selects to open the file.

**Comment by YaLTeR** [comment]
This is probably related: I'm trying to implement "Show in Files" for a file that my application saves using a `GtkFileChooserNative`. It seems that the `OpenURI.OpenDirectory` method returns an error for fds under `/run/user/1000/doc/...` that I get from `GtkFileChooserNative` or from `flatpak run --file-forwarding`. Shouldn't the portal do the necessary path conversion backwards in this case?

**Comment by YaLTeR** [comment]
I think the error part got fixed in https://github.com/flatpak/xdg-desktop-portal/pull/549 and the path conversion part is handled by the PR I just opened.

**Comment by Mikenux** [comment]
Note that these actions are on a notification. So shouldn't this be part of standard actions for notifications? @jonas2515?

**Comment by GeorgesStavracas** [comment]
Seems like this issue was fixed by the combination of https://github.com/flatpak/xdg-desktop-portal/pull/549, https://github.com/flatpak/xdg-desktop-portal/pull/552, https://github.com/flatpak/xdg-desktop-portal/pull/672.

Please reopen if there's anything else to do here!
